# Chapter 17

In [1]:
import pandas as pd
import statsmodels.api as sm
# import numpy as np
# import matplotlib.pyplot as plt

In [2]:
# Exercise 1
pntsprd = pd.read_stata("./stata/PNTSPRD.DTA")
X = sm.add_constant(pntsprd[["spread"]])
model = sm.OLS(pntsprd.favwin, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 favwin   R-squared:                       0.111
Model:                            OLS   Adj. R-squared:                  0.109
Method:                 Least Squares   F-statistic:                     68.57
Date:                Tue, 26 Jul 2022   Prob (F-statistic):           9.32e-16
Time:                        01:40:03   Log-Likelihood:                -279.29
No. Observations:                 553   AIC:                             562.6
Df Residuals:                     551   BIC:                             571.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.5769      0.028     20.434      0.000       0.521       0.632
spread         0.0194      0.002      8.281      0.000       0.015       0.024
==============================================================================
Omnibus:                       86.055   Durbin-Watson:                   2.112
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               94.402
Skew:                          -0.956   Prob(JB):                     3.17e-21
Kurtosis:                       2.336   Cond. No.                         20.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [3]:
model.t_test("const = 0.5")

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
c0             0.5769      0.028      2.725      0.007       0.521       0.632

In [4]:
model = sm.OLS(pntsprd.favwin, X, missing="drop").fit(cov_type="HC3")
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 favwin   R-squared:                       0.111
Model:                            OLS   Adj. R-squared:                  0.109
Method:                 Least Squares   F-statistic:                     100.8
Date:                Tue, 26 Jul 2022   Prob (F-statistic):           6.80e-22
Time:                        01:40:03   Log-Likelihood:                -279.29
No. Observations:                 553   AIC:                             562.6
Df Residuals:                     551   BIC:                             571.2
Df Model:                           1                                         
Covariance Type:                  HC3                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.5769      0.032     18.190      0.000       0.515       0.639
spread         0.0194      0.002     10.040      0.000       0.016       0.023
==============================================================================
Omnibus:                       86.055   Durbin-Watson:                   2.112
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               94.402
Skew:                          -0.956   Prob(JB):                     3.17e-21
Kurtosis:                       2.336   Cond. No.                         20.0
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC3)
"""

In [5]:
model.predict([1, 10])

array([0.77060438])

In [6]:
model.t_test("const = 0.5")

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
c0             0.5769      0.032      2.426      0.015       0.515       0.639

In [7]:
model = sm.Probit(pntsprd.favwin, X, missing="drop").fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.476604
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Probit Regression Results                           
==============================================================================
Dep. Variable:                 favwin   No. Observations:                  553
Model:                         Probit   Df Residuals:                      551
Method:                           MLE   Df Model:                            1
Date:                Tue, 26 Jul 2022   Pseudo R-squ.:                  0.1294
Time:                        01:40:03   Log-Likelihood:                -263.56
converged:                       True   LL-Null:                       -302.75
Covariance Type:            nonrobust   LLR p-value:                 8.521e-19
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0106      0.104     -0.102      0.919      -0.214       0.193
spread         0.0925      0.012      7.591      0.000       0.069       0.116
==============================================================================
"""

In [8]:
model.t_test("const = 0")

<class 'statsmodels.stats.contrast.ContrastResults'>
                             Test for Constraints                             
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
c0            -0.0106      0.104     -0.102      0.919      -0.214       0.193

In [9]:
model.predict([1, 10])

array([0.81965134])

In [10]:
X = sm.add_constant(pntsprd[["spread", "favhome", "fav25", "und25"]])
model = sm.Probit(pntsprd.favwin, X, missing="drop").fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.474940
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Probit Regression Results                           
==============================================================================
Dep. Variable:                 favwin   No. Observations:                  553
Model:                         Probit   Df Residuals:                      548
Method:                           MLE   Df Model:                            4
Date:                Tue, 26 Jul 2022   Pseudo R-squ.:                  0.1325
Time:                        01:40:03   Log-Likelihood:                -262.64
converged:                       True   LL-Null:                       -302.75
Covariance Type:            nonrobust   LLR p-value:                 1.567e-16
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0552      0.129     -0.429      0.668      -0.308       0.197
spread         0.0879      0.013      6.787      0.000       0.063       0.113
favhome        0.1486      0.137      1.084      0.278      -0.120       0.417
fav25          0.0031      0.159      0.019      0.985      -0.308       0.314
und25         -0.2198      0.251     -0.877      0.380      -0.711       0.271
==============================================================================
"""

In [11]:
model.wald_test("favhome = fav25 = und25 = 0", scalar=True) # Not LR but seems to work?

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=1.8440186856438163, p-value=0.605403142010621, df_denom=3>

C1.i When the spread is zero neither team is favoured to win, and so it should follow that the probability is even (50%).

C1.ii We reject the null hypothesis for both the ordinary and HC3 errors, at least at the 5% level (1% for the unadjusted errors, but given that it is the linear probability model, it is more appropriate to consider the robust errors). In either case, we reject the null that the constant is 0.5.

C1.iii Spread is statistically singificant under both sets of errors. The estimated probability under the linear probability model when spread = 10 is approximately 77%.

C1.iv We do not reject the null hypothesis that the constant is zero. This is the equivalent to our earlier test under the linear probability model, but takes into account that we are running a probit model. If the spread is 0 and the constant is 0 then we have the CDF of the normal distribution at 0, or, 0.5.

C1.v The probability under the probit model is a little higher at 82%

C1.vi The new variables are not jointly significant. The other factors do not anything to predicting the outcome once spread is accounted for.

In [12]:
# Exercise 2
loanapp = pd.read_stata("./stata/loanapp.dta")
X = sm.add_constant(loanapp[["white"]])
model = sm.OLS(loanapp.approve, X, missing="drop").fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                approve   R-squared:                       0.049
Model:                            OLS   Adj. R-squared:                  0.048
Method:                 Least Squares   F-statistic:                     102.2
Date:                Tue, 26 Jul 2022   Prob (F-statistic):           1.81e-23
Time:                        01:40:03   Log-Likelihood:                -555.54
No. Observations:                1989   AIC:                             1115.
Df Residuals:                    1987   BIC:                             1126.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.7078      0.018     38.806      0.000       0.672       0.744
white          0.2006      0.020     10.111      0.000       0.162       0.240
==============================================================================
Omnibus:                      801.085   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2394.523
Skew:                          -2.161   Prob(JB):                         0.00
Kurtosis:                       6.197   Cond. No.                         4.90
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [13]:
print(model.predict([1, 1]))
print(model.predict([1, 0]))

[0.90838786]
[0.70779221]


In [14]:
model = sm.Probit(loanapp.approve, X, missing="drop").fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.352377
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Probit Regression Results                           
==============================================================================
Dep. Variable:                approve   No. Observations:                 1989
Model:                         Probit   Df Residuals:                     1987
Method:                           MLE   Df Model:                            1
Date:                Tue, 26 Jul 2022   Pseudo R-squ.:                 0.05331
Time:                        01:40:04   Log-Likelihood:                -700.88
converged:                       True   LL-Null:                       -740.35
Covariance Type:            nonrobust   LLR p-value:                 6.408e-19
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.5469      0.075      7.251      0.000       0.399       0.695
white          0.7839      0.087      9.041      0.000       0.614       0.954
==============================================================================
"""

In [15]:
print(model.predict([1, 1]))
print(model.predict([1, 0]))

[0.90838786]
[0.70779221]


In [16]:
X = sm.add_constant(loanapp[["white", "hrat", "obrat", "loanprc", "unem", "male", "married", "dep", "sch", "cosign", "chist", "pubrec", "mortlat1", "mortlat2", "vr"]])
model = sm.Probit(loanapp.approve, X, missing="drop").fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.304551
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                          Probit Regression Results                           
==============================================================================
Dep. Variable:                approve   No. Observations:                 1971
Model:                         Probit   Df Residuals:                     1955
Method:                           MLE   Df Model:                           15
Date:                Tue, 26 Jul 2022   Pseudo R-squ.:                  0.1866
Time:                        01:40:04   Log-Likelihood:                -600.27
converged:                       True   LL-Null:                       -737.98
Covariance Type:            nonrobust   LLR p-value:                 7.014e-50
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.0623      0.313      6.585      0.000       1.449       2.676
white          0.5203      0.097      5.366      0.000       0.330       0.710
hrat           0.0079      0.007      1.131      0.258      -0.006       0.022
obrat         -0.0277      0.006     -4.578      0.000      -0.040      -0.016
loanprc       -1.0120      0.237     -4.266      0.000      -1.477      -0.547
unem          -0.0367      0.017     -2.099      0.036      -0.071      -0.002
male          -0.0370      0.110     -0.337      0.736      -0.252       0.178
married        0.2657      0.094      2.820      0.005       0.081       0.450
dep           -0.0496      0.039     -1.269      0.204      -0.126       0.027
sch            0.0146      0.096      0.153      0.879      -0.173       0.202
cosign         0.0861      0.246      0.350      0.726      -0.396       0.568
chist          0.5853      0.096      6.098      0.000       0.397       0.773
pubrec        -0.7787      0.126     -6.165      0.000      -1.026      -0.531
mortlat1      -0.1876      0.253     -0.741      0.459      -0.684       0.308
mortlat2      -0.4944      0.327     -1.514      0.130      -1.134       0.146
vr            -0.2011      0.081     -2.467      0.014      -0.361      -0.041
==============================================================================
"""

In [17]:
X_pred = X.dropna()
X_pred["white"] = 0
prob_avg = model.predict(X_pred).mean()
X_pred["white"] = 1
model.predict(X_pred).mean() - prob_avg

/tmp/ipykernel_6983/89657284.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_pred["white"] = 0
/tmp/ipykernel_6983/89657284.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_pred["white"] = 1


0.104224483076233

In [18]:
model = sm.Logit(loanapp.approve, X, missing="drop").fit()
model.summary()

Optimization terminated successfully.
         Current function value: 0.304666
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                approve   No. Observations:                 1971
Model:                          Logit   Df Residuals:                     1955
Method:                           MLE   Df Model:                           15
Date:                Tue, 26 Jul 2022   Pseudo R-squ.:                  0.1863
Time:                        01:40:05   Log-Likelihood:                -600.50
converged:                       True   LL-Null:                       -737.98
Covariance Type:            nonrobust   LLR p-value:                 8.693e-50
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.8017      0.595      6.393      0.000       2.636       4.967
white          0.9378      0.173      5.424      0.000       0.599       1.277
hrat           0.0133      0.013      1.030      0.303      -0.012       0.039
obrat         -0.0530      0.011     -4.701      0.000      -0.075      -0.031
loanprc       -1.9050      0.460     -4.137      0.000      -2.807      -1.002
unem          -0.0666      0.033     -2.029      0.042      -0.131      -0.002
male          -0.0664      0.206     -0.322      0.748      -0.471       0.338
married        0.5033      0.178      2.827      0.005       0.154       0.852
dep           -0.0907      0.073     -1.237      0.216      -0.234       0.053
sch            0.0412      0.178      0.231      0.817      -0.308       0.391
cosign         0.1321      0.446      0.296      0.767      -0.742       1.006
chist          1.0666      0.171      6.230      0.000       0.731       1.402
pubrec        -1.3407      0.217     -6.168      0.000      -1.767      -0.915
mortlat1      -0.3099      0.464     -0.669      0.504      -1.218       0.599
mortlat2      -0.8947      0.569     -1.574      0.116      -2.009       0.220
vr            -0.3498      0.154     -2.276      0.023      -0.651      -0.049
==============================================================================
"""

In [19]:
X_pred = X.dropna()
X_pred["white"] = 0
log_avg = model.predict(X_pred).mean()
X_pred["white"] = 1
model.predict(X_pred).mean() - log_avg

/tmp/ipykernel_6983/1197556308.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_pred["white"] = 0
/tmp/ipykernel_6983/1197556308.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_pred["white"] = 1


0.10087064751405628

C2.i Estimates above. Both the Probit and LPM are identical.

C2.ii White is positive and significant, suggesting there is evidence of discrimination.

C2.iii The coefficient on white is notably higher but still positive and significant.

C2.iv The size of the effect under probit is 10.4%, while the logit effect is about 10.1%